### PR TITLE
Optional secret for Assignable CapAccess

### PR DESCRIPTION
**Pull request by maackle**
### Summary

This came out of a documentation discussion. Doesn't this seem preferable?

Also looks like there was a bug, which this fixes

### TODO:
- [ ] CHANGELOGs updated with appropriate info

### CODE REVIEW

File: crates/holochain_zome_types/src/fixt.rs [hunk 1]
```diff
@@ -359,7 +359,7 @@ fixturator!(
             CapAccessVariant::Unrestricted => CapAccess::from(()),
             CapAccessVariant::Transferable => CapAccess::from(CapSecretFixturator::new_indexed(Empty, get_fixt_index!()).next().unwrap()),
             CapAccessVariant::Assigned => CapAccess::from((
-                CapSecretFixturator::new_indexed(Empty, get_fixt_index!()).next().unwrap(),
+                Some(CapSecretFixturator::new_indexed(Empty, get_fixt_index!()).next().unwrap()),
```
Comment (pdaoust) [inline crates/holochain_zome_types/src/fixt.rs:362]: Would it be more semantically correct to use `None` for an empty fixture? (I have no clue; just applying beginner's mind here)

File: crates/holochain_integrity_types/src/capability/grant.rs [hunk 2]
```diff
@@ -111,13 +111,19 @@ impl CapGrant {
                     // The grant has no assignees so is always valid…
                     _ => true,
                 }
-                // The secret needs to match…
-                && match access {
-                    // Unless the extern is unrestricted.
-                    CapAccess::Unrestricted => true,
-                    // note the PartialEq implementation is constant time for secrets
-                    CapAccess::Transferable { secret, .. } => check_secret.map(|given| secret == given).unwrap_or(false),
-                    CapAccess::Assigned { secret, .. } => check_secret.map(|given| secret == given).unwrap_or(false),
```
[the remark this answers is not in the document]
Comment (matthme) [reply crates/holochain_integrity_types/src/capability/grant.rs:120]: Well, if this function named `is_valid` is being called for a `Transferable` cap grant but with `check_secret == None` then the result cannot really be represented with a boolean in the first place, can it? It's rather an "undetermined" state - so maybe the function should be called `is_validated` instead (small joke ;-) )?
So I see how one could reason to return false if no secret is provided for a Transferable cap grant to protect against this function accidentally being called for a transferable grant without providing the secret.

File: crates/holochain_zome_types/src/fixt.rs [hunk 3]
```diff
@@ -376,7 +376,7 @@ fixturator!(
                 let number_of_assigned = rng.gen_range(0..5);
 
                 CapAccess::from((
-                    CapSecretFixturator::new_indexed(Unpredictable, get_fixt_index!()).next().unwrap(),
+                    Some(CapSecretFixturator::new_indexed(Unpredictable, get_fixt_index!()).next().unwrap()),
```
Comment (pdaoust) [inline crates/holochain_zome_types/src/fixt.rs:379]: Likewise, wondering if `None` would also be a valid point in the unpredictable curve.

**Review by pdaoust — DISMISSED** [review]

File: crates/holochain_integrity_types/src/capability/grant.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (pdaoust) [reply crates/holochain_integrity_types/src/capability/grant.rs:120]: @maackle after reading your comment in light of @matthme 's comment I think I see what you're saying. The effect of skipping the check would be that the loop that checks all the grants on the source chain would also come up with nothing and return false, right? I'm wondering when are the times that the secret would be `None` -- AFAICT it's only when the caller didn't supply one, in which case I think this check should probably return `false` rather than getting skipped (although the effect is the same -- access denied).

[the remark this answers is not in the document]
Comment (jost-s) [reply crates/holochain_integrity_types/src/capability/grant.rs:120]: > If check_secret is None, then this function would always just return false

For unrestricted cap grants it returns `true`, and the other two variants required a secret, so you couldn't skip the secret check. It seemed correct to me.

If the secret becomes optional for assigned cap grants, the logic can no longer be used. But the update is wrong, because a transferable cap grant still requires a secret, and now if none is provided, the secret check is skipped.

File: crates/holochain_integrity_types/src/capability/grant.rs [hunk 4]
```diff
@@ -111,13 +111,19 @@ impl CapGrant {
...
+                // The secret needs to match, if provided
+                && {
+                    if let Some(given_secret) = check_secret {
+                        match access {
+                            // Unless the extern is unrestricted.
+                            CapAccess::Unrestricted => true,
+                            // note the PartialEq implementation is constant time for secrets
+                            CapAccess::Transferable { secret, .. } => secret == given_secret,
```
Comment (jost-s) [inline crates/holochain_integrity_types/src/capability/grant.rs:121]: Transferable requires a secret, the check must not be skipped by not providing one.

**Review by jost-s — CHANGES_REQUESTED** [review]

File: crates/holochain_integrity_types/src/capability/grant.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (pdaoust) [reply crates/holochain_integrity_types/src/capability/grant.rs:120]: > this check should probably return false rather than getting skipped (although the effect is the same -- access denied).

In order to properly implement 'skipped' logic in a way that doesn't let [`SourceChain::valid_cap_grant`](https://github.com/holochain/holochain/blob/c745c93471d697cbbe2f1f3478a0a8dacbf00ffc/crates/holochain_state/src/source_chain.rs#L652-L657) allow unauthorised people access a function by supplying a `None` secret, this function would have to have a return signature of `Option<bool>` and let its consumer treat `None` as equivalent to `false` in the loop... (which I guess is what you're saying @matthme ) I'd just as soon return `false` if a required secret wasn't supplied.

File: crates/holochain_zome_types/src/fixt.rs (hunk 3 above)
Comment (maackle) [reply crates/holochain_zome_types/src/fixt.rs:379]: It would, but it's onerous to make an fixturator for an Option of something, so I opted to just replicate the previous behavior.

File: crates/holochain_zome_types/src/fixt.rs (hunk 1 above)
Comment (maackle) [reply crates/holochain_zome_types/src/fixt.rs:362]: Yeah, that would probably be better

File: crates/holochain_integrity_types/src/capability/grant.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (maackle) [reply crates/holochain_integrity_types/src/capability/grant.rs:120]: I made the mistake of interpreting `check_secret` as an imperative i.e. if it's Some, we want to check the secret, if it's None, we don't want to check the secret, and it's the responsibility of the caller to decide whether they want to check the secret or not. I did this too casually, looking more deeply I see that check_secret comes directly from the zome caller, so we can't give them that responsibility.

Let me try again.

File: crates/holochain_zome_types/src/fixt.rs (hunk 3 above)
Comment (pdaoust) [reply crates/holochain_zome_types/src/fixt.rs:379]: got it. Even knowing nothing about how fixt works, I kinda anticipated that it would be complicated.

**Review by pdaoust — COMMENTED** [review]
Some positive feedback and a non-blocking suggestion.

File: crates/holochain_integrity_types/src/capability/grant.rs [hunk 5]
```diff
@@ -87,37 +87,43 @@ impl CapGrant {
...
+
                     // note the PartialEq implementation is constant time for secrets
-                    CapAccess::Transferable { secret, .. } => check_secret.map(|given| secret == given).unwrap_or(false),
-                    CapAccess::Assigned { secret, .. } => check_secret.map(|given| secret == given).unwrap_or(false),
+                    // so this is safe from timing attacks
+
+                    // Transferable grants always require a matching secret
+                    CapAccess::Transferable { secret, .. } => Some(secret) == given_secret,
+                    // If a secret is required for the Assigned grant, a matching secret must be provided
+                    CapAccess::Assigned { secret: Some(secret), .. } => given_secret.map(|given| secret == given).unwrap_or(false)
```
[the remark this answers is not in the document]
Comment (maackle) [reply crates/holochain_integrity_types/src/capability/grant.rs:126]: I didn't do that because that would have the effect of penalizing someone for providing a secret for a cap that doesn't require one. I wanted it so that if the Assigned cap grant has a None secret, it's valid to provide a secret or not.

File: crates/holochain_integrity_types/src/capability/grant.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jost-s) [reply crates/holochain_integrity_types/src/capability/grant.rs:120]: LGTM. It would be great to have tests for this method. Are you willing to add them? Otherwise I would.

File: crates/holochain_integrity_types/src/capability/grant.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (pdaoust) [reply crates/holochain_integrity_types/src/capability/grant.rs:126]: oh ha ha, you're right!

File: crates/holochain_integrity_types/src/capability/grant.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (maackle) [reply crates/holochain_integrity_types/src/capability/grant.rs:120]: Test written.

File: crates/holochain_integrity_types/src/capability/grant.rs [hunk 6]
```diff
@@ -204,3 +210,116 @@ pub enum GrantedFunctions {
...
+    #[test]
+    fn cap_grant_is_valid() {
+        let agent1 = AgentPubKey::from_raw_36(vec![1; 36]);
+        let agent2 = AgentPubKey::from_raw_36(vec![2; 36]);
+        let assignees: BTreeSet<_> = [agent1.clone()].into_iter().collect();
+        let secret: CapSecret = [1; 64].into();
+        let secret_wrong: CapSecret = [2; 64].into();
+        let tag = "tag".to_string();
+
+        let g1: CapGrant = ZomeCallCapGrant {
```
Comment (jost-s) [inline crates/holochain_integrity_types/src/capability/grant.rs:226]:
```suggestion
        let transferable: CapGrant = ZomeCallCapGrant {
```

File: crates/holochain_integrity_types/src/capability/grant.rs [hunk 7]
```diff
@@ -204,3 +210,116 @@ pub enum GrantedFunctions {
...
+            tag: tag.clone(),
+            access: CapAccess::Transferable {
+                secret: secret.clone(),
+            },
+            functions: GrantedFunctions::All,
+        }
+        .into();
+        let g2: CapGrant = ZomeCallCapGrant {
```
Comment (jost-s) [inline crates/holochain_integrity_types/src/capability/grant.rs:234]:
```suggestion
        let assigned_to_agent_1_no_secret: CapGrant = ZomeCallCapGrant {
```

File: crates/holochain_integrity_types/src/capability/grant.rs [hunk 8]
```diff
@@ -204,3 +210,116 @@ pub enum GrantedFunctions {
...
+            tag: tag.clone(),
+            access: CapAccess::Assigned {
+                secret: None,
+                assignees: assignees.clone(),
+            },
+            functions: GrantedFunctions::All,
+        }
+        .into();
+        let g3: CapGrant = ZomeCallCapGrant {
```
Comment (jost-s) [inline crates/holochain_integrity_types/src/capability/grant.rs:243]:
```suggestion
        let assigned_to_agent_1_with_secret: CapGrant = ZomeCallCapGrant {
```

**Review by jost-s — APPROVED** [review]
Yay!

It makes me doubt that passing in a wrong secret when none is required is valid. Fine anyway.

**Comment by maackle** [comment]
@pdaoust want to re-review?

**Comment by maackle** [comment]
I stand corrected, this is not preferable:

- the "secret" is the access token, and without that we are implementing an Access Control List, which we are trying to avoid
- the only easily queryable field in the database is the `secret` field, so without that provided, we'd have to design a new, less-efficient query

**Comment by jost-s** [comment]
I would like for you to merge the tests you've written. They are valuable IMO.

**Comment by maackle** [comment]
@jost good point, the test is on #3961
